url5e6a51063b_nuigroup_hci_for_java_tgz-pJ8-nui_squirt_demo_DemoSquirtApplicationJ8/src/nui/squirt/component/TextField.java

```diff
@@ -0,0 +1,347 @@
+						}
+						else if (i > s) {
+							setSecondaryCaretIndex(i);
+						}
+						else if (i < s) {
+							setCaretIndex(s);
+						}
+					}
+					break;
+				case KeyEvent.SHIFT:
+					shiftDown = true;
+					break;
+			}
+		}
+		timer = 0;
+	}
+
+	public void keyReleased(KeyEvent e) {
+		if (e.isCoded()) {
+			switch (e.getKeyCode()) {
+				case KeyEvent.SHIFT:
+					shiftDown = false;
+					break;
+			}
+		}
+	}
+
+	public void keyTyped(KeyEvent e) {
+		if (!e.isCoded()) {
+			if (e.getKey() == KeyEvent.BACKSPACE && getCaretIndex() > 0) {
+				setText(getText().substring(0, getCaretIndex()-1) + getText().substring(getCaretIndex()));
+				setCaretIndex(getCaretIndex()-1);
+			}
+			else if (e.getKey() == KeyEvent.DELETE && getCaretIndex() < getText().length()) {
+				setText(getText().substring(0, getCaretIndex()) + getText().substring(getCaretIndex()+1));
+			}
+			else if (e.getKey() != KeyEvent.DELETE && e.getKey() != KeyEvent.BACKSPACE) {
+				int min = Math.min(getCaretIndex(), getSecondaryCaretIndex());
+				int max = Math.max(getCaretIndex(), getSecondaryCaretIndex());
+				setText(getText().substring(0, min) + e.getKey() + getText().substring(max));
+				setCaretIndex(min+1);
+			}
+			setSecondaryCaretIndex(getCaretIndex());
+		}
+	}
+
+	public void addTextListener(TextListener l) {
+		listeners.add(l);
+	}
+
+	public void removeTextListener(TextListener l) {
+		listeners.remove(l);
+	}
+
+	public void fireTextChanged(TextEvent e) {
+		for (TextListener l: listeners) {
+			l.textChanged(e);
+		}
+	}
+
+	public void addKeyboard(Keyboard k) {
+		keyboards.add(k);
+	}
+
+	public void removeKeyboard(Keyboard k) {
+		keyboards.remove(k);
+	}
+	
+	@Override
+	public boolean canAcceptMoreControlPoints() {
+		return controlPoint == null;
+	}
+	
+	@Override
+	public boolean offer(ControlPoint cp) {
+		cp.addControlPointListener(this);
+		return true;
+	}
+	
+	@Override
+	public void controlPointCreated(ControlPoint cp) {
+		if (controlPoint == null && isUnderPoint(cp)) {
+			this.controlPoint = cp;
+			NUIController.getInstance().addKeyListener(this);
+			PVector local = transformToLocalSpace(new PVector(cp.getX(), cp.getY()));
+			local.add(getWidth()/2 - TEXT_FIELD_TEXT_PADDING, getHeight()/2 - TEXT_FIELD_TEXT_PADDING, 0);
+			setCaretIndex(mapPosToStringIndex(local));
+			if (!shiftDown) {
+				setSecondaryCaretIndex(getCaretIndex());
+			}
+		}
+	}
+	
+	@Override
+	public void controlPointDied(ControlPoint cp) {
+		if (controlPoint != null && controlPoint.equals(cp)) {
+			controlPoint = null;
+		}
+	}
+	
+	@Override
+	public void controlPointUpdated(ControlPoint cp) {
+		if (controlPoint != null && controlPoint.equals(cp)) {
+			PVector local = transformToLocalSpace(new PVector(cp.getX(), cp.getY()));
+			local.add(getWidth()/2 - TEXT_FIELD_TEXT_PADDING, getHeight()/2 - TEXT_FIELD_TEXT_PADDING, 0);
+			setCaretIndex(mapPosToStringIndex(local));
+		}
+	}
+
+	private int mapPosToStringIndex(PVector pos) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 14. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 327-336 into a separate method.